src/i256.rs

```diff
@@ -0,0 +1,279 @@
+        let rhs = if rhs_sign == 0 { rhs } else { -rhs };
+
+        // Mustn't overflow because we're usually promoting just i128 to I256.
+        let result = Self::new(lhs.inner * rhs.inner);
+        if lhs_sign ^ rhs_sign == 0 {
+            result
+        } else {
+            -result
+        }
+    }
+}
+
+impl Div for I256 {
+    type Output = Self;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        let lhs_sign = self.sign();
+        let rhs_sign = rhs.sign();
+
+        let lhs = if lhs_sign == 0 { self } else { -self };
+        let rhs = if rhs_sign == 0 { rhs } else { -rhs };
+
+        // Mustn't overflow because we're usually promoting just i128 to I256.
+        let result = Self::new(lhs.inner / rhs.inner);
+        if lhs_sign ^ rhs_sign == 0 {
+            result
+        } else {
+            -result
+        }
+    }
+}
+
+impl Sub for I256 {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        let lhs_sign = self.sign();
+        let rhs_sign = rhs.sign();
+
+        let lhs = if lhs_sign == 0 { self } else { -self };
+        let rhs = if rhs_sign == 0 { rhs } else { -rhs };
+
+        let result = Self::new(lhs.inner - rhs.inner);
+        if lhs_sign ^ rhs_sign == 0 {
+            result
+        } else {
+            -result
+        }
+    }
+}
+
+impl Neg for I256 {
+    type Output = Self;
+
+    /// N.B. Neg has a single case of panicking: `-I256::MIN`
+    /// Because on two's complement we always have one extra negative value
+    fn neg(self) -> Self::Output {
+        if self == Self::MIN {
+            panic_on_overflow();
+        }
+        const U1: U256 = I256::from_i128(1).inner;
+        // Overflow takes place when we negate zero.
+        let (x, _) = (!self.inner).overflowing_add(U1);
+        Self::new(x)
+    }
+}
+
+impl cmp::Ord for I256 {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        match (self.is_negative(), other.is_negative()) {
+            (true, false) => cmp::Ordering::Less,
+            (false, true) => cmp::Ordering::Greater,
+            _ => self.inner.cmp(&other.inner),
+        }
+    }
+}
+
+impl cmp::PartialOrd for I256 {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl From<i128> for I256 {
+    fn from(x: i128) -> Self {
+        Self::from_i128(x)
+    }
+}
+
+impl TryFrom<I256> for i128 {
+    type Error = ArithmeticError;
+
+    fn try_from(x: I256) -> Result<Self, Self::Error> {
+        if x > I256::I128_MAX || x < I256::I128_MIN {
+            return Err(ArithmeticError::Overflow);
+        }
+        if x == I256::I128_MIN {
+            return Ok(i128::MIN);
+        }
+        let was_negative = x.is_negative();
+        let x: i128 = x
+            .abs()
+            .inner
+            .try_into()
+            .map_err(|_| ArithmeticError::Overflow)?;
+        Ok(if was_negative { -x } else { x })
```

Comment:
Why not just `u128::from(w[0]) && u128::from(w[1]) << 64` after the first check?

Reply:
done